src/transformers/mocha.test.ts

```diff
@@ -162,3 +162,79 @@ describe('suite', () => {
 })
     `
 )
+
+testChanged(
+  'transforms this',
+  `
+// @flow
+describe('describe', function () {
+  before(function() {
+    this.hello = 'hi';
+  });
+  afterAll(function () {
+    console.log(this.hello);
+  });
+  beforeEach(function () {
+    this.goodbye = 'bye';
+  });
+  afterEach(function () {
+    console.log(this.hello);
+  });
+
+  before('some text', () => {});
+  after('some text', () => {});
+  beforeEach('some text', () => {});
+  afterEach('some text', () => {});
+
+  context('context', () => {
+    it('it', function () {
+      console.log(this.hello);
+      console.log(this.goodbye);
+    });
+    specify('specify', function () {
+      console.log(this.hello);
+      console.log(this.goodbye);
+    })
+  })
+})
+`,
+  `
+// @flow
+describe('describe', () => {
```

Comment:
I've formatted this using the codemod, so I believe it should be:
```
// @flow
describe('describe', () => {
  let testContext;

  beforeEach(() => {
    testContext = {};
  });

  beforeAll(function() {
    testContext.hello = 'hi';
  });
  afterAll(function () {
    console.log(testContext.hello);
  });
  beforeEach(() => {
    testContext.goodbye = 'bye';
  });
  afterEach(() => {
    console.log(testContext.hello);
  });
  beforeAll(() => {});
  afterAll(() => {});
  beforeEach(() => {});
  afterEach(() => {});
  describe('context', () => {
    test('it', () => {
      console.log(testContext.hello);
      console.log(testContext.goodbye);
    });
    test('specify', () => {
      console.log(testContext.hello);
      console.log(testContext.goodbye);
    })
  })
})

```

Reply:
I believe this is invalid because from what I can infer from the docs (https://jestjs.io/docs/en/setup-teardown#scoping) and basic testing the `beforeAll` will execute before the `beforeEach` so this will error with: `TypeError: Cannot set property 'hello' of undefined` since `testContext` hasn't been initialized when `testContext.hello = 'hi';` is invoked.

### Invalid

```typescript
describe('example', () => {
  let testContext

  beforeEach(() => {
    testContext = {}
  })

  beforeAll(() => {
    testContext.hello = 'hi'
  })

  it(`case 1`, () => {
    console.log(testContext)
  })

  it(`case 2`, () => {
    console.log(testContext)
  })
})
```

This errors:

```
TypeError: Cannot set property 'hello' of undefined
TypeError: Cannot set property 'hello' of undefined
```

### Valid

```typescript
describe('example', () => {
  let testContext

  beforeAll(() => {
    testContext = {}
  })

  beforeAll(() => {
    testContext.hello = 'hi'
  })

  it(`case 1`, () => {
    console.log(testContext)
  })

  it(`case 2`, () => {
    console.log(testContext)
  })
})
```

This prints:

```
  console.log
    { hello: 'hi' }
  console.log
    { hello: 'hi' }
```

My thinking was to have a "ranking" of these global hooks. If there is a `beforeAll` that uses `this`, then the corresponding setup hook (`testContext = {}`) would be in a `beforeAll`. If there isn't a `beforeAll`, but there is a `beforeEach`, then use a `beforeEach`.

Reply:
That makes sense. I was confused by the `before` function here (https://github.com/skovhus/jest-codemods/pull/182/files/e1961b08653476fc9145f3d9f4453e28a3537ea6#r335165923)